nsiqcppstyle_rulemanager.py

```diff
@@ -172,77 +192,93 @@ def RunProjectRules(self, targetName):
 
     def ResetRegisteredRules(self):
         """ Reset all registered rules. """
-        del self.functionNameRules[:]
-        del self.functionScopeRules[:]
-        del self.lineRules[:]
-        del self.rules[:]
-        del self.typeNameRules[:]
-        del self.typeScopeRules[:]
-        del self.fileStartRules[:]
-        del self.fileEndRules[:]
-        del self.projectRules[:]
-        del self.preprocessRules[:]
-
-    def AddPreprocessRule(self, preprocessRule):
+
+        self.functionNameRules.clear()
+        self.functionScopeRules.clear()
+        self.lineRules.clear()
+        self.rules.clear()
+        self.typeNameRules.clear()
+        self.typeScopeRules.clear()
+        self.fileStartRules.clear()
+        self.fileEndRules.clear()
+        self.sessionStartRules.clear()
+        self.sessionEndRules.clear()
+        self.projectRules.clear()
+        self.preprocessRules.clear()
+        self.commentRules.clear()
+
+    def AddPreprocessRule(self, user_function: Callable[[LexerType, ContextStackType], None]):
         """ Add rule which runs in preprocess statements """
-        self.preprocessRules.append(preprocessRule)
+        self.preprocessRules.append(user_function)
+
+    def AddCommentRule(self, user_function: Callable[[LexerType, TokenType], None]):
+        """ Add rule which runs when a comment is encountered """
+        self.commentRules.append(user_function)
 
-    def AddFunctionScopeRule(self, functionScopeRule):
+    def AddFunctionScopeRule(self, user_function: Callable[[LexerType, ContextStackType], None]):
         """ Add rule which runs in function scope """
-        self.functionScopeRules.append(functionScopeRule)
+        self.functionScopeRules.append(user_function)
 
-    def AddFunctionNameRule(self, functionRule):
-        """ Add rule on the function name place"""
-        self.functionNameRules.append(functionRule)
+    def AddFunctionNameRule(self,
+                            user_function: Callable[
+                                [LexerType, FullNameType, DeclarationType, ContextStackType, ContextType],
+                                None]):
+        """ Add rule on the function name place """
+        self.functionNameRules.append(user_function)
 
-    def AddLineRule(self, lineRule):
+    def AddLineRule(self, user_function: Callable[[LexerType, LineType, LineNumberType], None]):

```

Comment:
While reading this, it occurs to me that the following is much more reader friendly:

```suggestion
    def AddLineRule(self, user_function: Callable[[Lexer, Line, LineNumber], None]):
```
It also follows the typing convention since `CallableType` is not used, but `Callable` is used in Python stdlib

Reply:
My fear was that some of the types would be too generic without the "Type" suffix (e.g., "Lexer", "Line").  I hope that adherence to naming conventions will leave these type names unique.

Done.